### PR TITLE
Implement JIT classes 

### PR DESCRIPTION
**Pull request by sklam**
* JIT class (as described in #1433 )
* Partial immutable jit class (only for nopython mode)
* fixes to optional types attribute access
* many fixes to casting and stricter type checking in it.
* introduce deferred type for recursive type in JIT classes
* examples of jit classes (singley linkedlist, stack, binary tree)

### CODE REVIEW

**Comment by seibert** [comment]
With the massive NBEP2 refactor, this PR is now in conflict with master.

File: numba/jitclass/base.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        outer = self
+
+        class ConstructorTemplate(templates.AbstractTemplate):
+            key = outer.class_type
+
+            def generic(self, args, kws):
+                ctor = outer.jitmethods['__init__']
+
+                instance_type = outer.class_type.instance_type
+                if instance_type not in defined_types:
```
Comment (pitrou) [inline numba/jitclass/base.py]: Is there any reason this is done lazily?

File: numba/jitclass/immutable.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+class ImmutableClassBuilder(ClassBuilder):
+    instance_type_class = types.ImmutableClassInstanceType
+
+    def implement_frontend(self, instance_type):
+        self.implement_attribute_typing(instance_type)
+        self.register_byref_attribute_typing(instance_type)
+
+    def register_byref_attribute_typing(self, instance_type):
+        class RefStructAttribute(templates.AttributeTemplate):
```
Comment (pitrou) [inline numba/jitclass/immutable.py]: If this is the same template for all reference types, why not define it at the top level with `ImmutableClassRefType` as the key? Or am I missing something?

File: numba/jitclass/immutable.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                if attr in structdct:
+                    return structdct[attr]
+
+        refattrspec = RefStructAttribute(self.typer,
+                                         instance_type.get_reference_type())
+        self.typer.insert_attributes(refattrspec)
+
+    def implement_constructor(self, registry, instance_type, ctor_nargs):
+        @registry.register
+        @imputils.implement(self.class_type, *([types.Any] * ctor_nargs))
```
Comment (pitrou) [inline numba/jitclass/immutable.py]: Same here with `ImmutableClassInstanceType` instead of `class_type`?

File: numba/runtime/nrt.c [hunk 4]
```diff
@@ -61,6 +61,14 @@ struct MemSys {
 /* The Memory System object */
 static NRT_MemSys TheMSys;
 
+
+typedef struct {
+    size_t allocated_size;
+    NRT_dtor_function dtor;
```
Comment (pitrou) [inline numba/runtime/nrt.c]: Why not change the `NRT_dtor_function` signature to take `(ptr, size, info)` parameters? This would simplify things a lot.

File: numba/targets/builtins.py [hunk 5]
```diff
@@ -511,6 +511,24 @@ def optional_is_not_none(context, builder, sig, args):
                   types.none, types.Optional)(optional_is_none))
 
 
+@builtin_attr
+@impl_attribute_generic(types.Optional)
+def optional_getattr(context, builder, typ, value, attr):
```
Comment (pitrou) [inline numba/targets/builtins.py]: This should go into the `optional` module, right?

File: numba/targets/imputils.py [hunk 6]
```diff
@@ -72,6 +72,23 @@ def imp(context, builder, sig, args):
             builder, func, fndesc.restype, fndesc.argtypes, args, env=None)
         with cgutils.if_unlikely(builder, status.is_error):
             context.call_conv.return_status_propagate(builder, status)
+        assert sig.return_type == fndesc.restype
+        # Reconstruct optional return type
+        if isinstance(sig.return_type, types.Optional):
```
Comment (pitrou) [inline numba/targets/imputils.py]: Can you explain why this is needed, while fndesc.restype is already an Optional?

File: numba/targets/base.py [hunk 7]
```diff
@@ -667,6 +713,29 @@ def cast(self, builder, val, fromty, toty):
         elif fromty == types.none and isinstance(toty, types.Optional):
             return self.make_optional_none(builder, toty.type)
 
+        elif (isinstance(toty, types.Optional) and
```
Comment (pitrou) [inline numba/targets/base.py]: In light of the various Optional enhancements, it would be nice to add tests to `test_optional`, IMHO.

File: examples/linkedlist.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    def prepend(self, data):
+        return LinkedNode(data, self)
+
+
+@njit
+def make_linked_node(data):
+    return LinkedNode(data, None)
+
+
+node_type.define(LinkedNode.class_type.instance_type)
```
Comment (pitrou) [inline examples/linkedlist.py]: While the recursive data structure examples are pleasant, I'm a bit skeptical about the deffered type design. The deferred type ends up leaking in all aspects of typing and code generation, and we have to special-case some operations for it (as is already the case with Optional).

Perhaps all deferred references could be resolved at type instantiation, so for example instead of writing:
```node_type.define(LinkedNode.class_type.instance_type)```
you would write :
``node_type = node_type.resolve(LinkedNode.class_type.instance_type)```
?

Comment (pitrou) [inline examples/linkedlist.py]: Another way to look at it would be to define temporary local names used for resolution of the class type:

```
spec = OrderedDict()

spec['data'] = int32
spec['next'] = optional(deferred('LinkedNode'))

@jitclass(spec)
class LinkedNode: # etc.

```

If the `jitclass` decorator is lazy enough, this could even be made to work for mutually recursive structures.

File: numba/jitclass/base.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+
+        self.typer.insert_function(ConstructorTemplate(self.typer))
+
+    def implement_frontend(self, instance_type):
+        self.implement_attribute_typing(instance_type)
+
+    def implement_attribute_typing(self, instance_type):
+        outer = self
+
+        class ClassAttribute(templates.AttributeTemplate):
```
Comment (pitrou) [inline numba/jitclass/base.py]: Same questions here as in the `immutable` module :-)

File: numba/jitclass/base.py [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+    builder: the internal jitclass builder
+    """
+    # Normalize spec
+    if isinstance(spec, Sequence):
+        spec = OrderedDict(spec)
+
+    # Copy methods from base classes
+    clsdct = {}
+    for basecls in reversed(inspect.getmro(cls)):
+        clsdct.update(basecls.__dict__)
```
Comment (pitrou) [inline numba/jitclass/base.py]: It seems this may also copy unrelated stuff such as `__module__`.

File: numba/jitclass/boxing.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                                                                lluintp))
+    addr_dataptr = c.pyapi.from_native_value(types.uintp,
+                                             c.builder.ptrtoint(dataptr,
+                                                                lluintp))
+
+    # XXX: relies on runtime address
+    int_addr_typ = c.context.get_constant(types.uintp, id(typ))
+
+    int_addr_boxcls = c.context.get_constant(types.uintp,
+                                             id(Box))
```
Comment (pitrou) [inline numba/jitclass/boxing.py]: You could create _specialize_box() here, no? Which would remove the need for a custom `__new__` and make boxing somewhat less slow...

File: numba/jitclass/boxing.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    """
+    __slots__ = '_meminfo', '_meminfoptr', '_dataptr', '_numba_type_'
+
+    def __new__(cls, meminfoptr, dataptr, typ):
+        if cls is not Box:
+            raise TypeError("cannot extend from specialized box")
+        return object.__new__(_specialize_box(typ))
+
+    def __init__(self, meminfoptr, dataptr, typ):
+        self._meminfo = MemInfo(meminfoptr)
```
Comment (pitrou) [inline numba/jitclass/boxing.py]: If `_meminfo` is only used for debug, it could be made a dynamic property (to reduce execution time of `__init__`).

File: numba/jitclass/base.py (hunk 1 above)
Comment (sklam) [inline numba/jitclass/base.py]: No.  It's due to earlier version of allowing some crazy way of specializing the class type.  Since that is removed, we don't need to lazily do this anymore.  Thanks for catching this.

File: numba/jitclass/immutable.py (hunk 2 above)
Comment (sklam) [inline numba/jitclass/immutable.py]: We can but it only works for getattr.  We need to fix it for setattr at the typing context first.

File: numba/runtime/nrt.c (hunk 4 above)
Comment (sklam) [inline numba/runtime/nrt.c]: Yes, indeed.

File: numba/targets/builtins.py (hunk 5 above)
Comment (sklam) [inline numba/targets/builtins.py]: `numba.targets.optional` only contains helpers for optional types while the rest of the implementation of optional are in `numba.targetsbuiltins` just before these lines.  We can move them in a separate PR.

File: numba/targets/imputils.py (hunk 6 above)
Comment (sklam) [inline numba/targets/imputils.py]: numba-compiled function unpacks optional return type and uses return-code to indicate whether it returns None or the value.  As a result the LLVM value of the returned value is actually not a optional type.  It is legacy code before the introduction of the ``Optional``.  Perhaps, we should unify everything to use Optional and remove the return-code for None.

File: numba/jitclass/boxing.py (hunk 12 above)
Comment (sklam) [inline numba/jitclass/boxing.py]: It is relying on `MemInfo` for owning the pointer and release it when it goes away.  I guess this deserves some comments.

File: numba/targets/imputils.py (hunk 6 above)
Comment (pitrou) [inline numba/targets/imputils.py]: Ah, right. Well, unifying would sound reasonable. A later PR I suppose?

File: examples/linkedlist.py (hunk 8 above)
Comment (sklam) [inline examples/linkedlist.py]: While I do think the deferred_type API can be improved, I do not understand several things from your suggestions:

1) Can you elaborate on "leaking in all aspects of typing and code generation".  Are you referring to the cyclic reference?  

2) I don't understand how `node_type = node_type.resolve(LinkedNode.class_type.instance_type)` would work.  

3) Does `deferred('LinkedNode')` require registering the name "LinkedNode" into the type system?  This seems to make the typesystem more complicated.  There will be name conflict problem to solve as well.

Comment (pitrou) [inline examples/linkedlist.py]: > Can you elaborate on "leaking in all aspects of typing and code generation". Are you referring to the cyclic reference?

No, I mean handling the deferred type adds some special cases at various points along the compilation pipeline (because the deferred type is semantically equivalent to the "real" type, but the type system doesn't allow recognizing that automatically -- and I have no idea how we could change that).

> 2) I don't understand how node_type = node_type.resolve(LinkedNode.class_type.instance_type) would work. 

Yes, sorry, my suggestion was wrong. My bad.

> Does deferred('LinkedNode') require registering the name "LinkedNode" into the type system?

No, I think it could work simply as a local name for resolution when a jitclass is compiled. Perhaps I'm missing something, though.
I agree registering the name into the type system would add a layer of complexity that's probably not welcome.

Comment (sklam) [inline examples/linkedlist.py]: > I think it could work simply as a local name for resolution when a jitclass is compiled.

The way I imagined the resolution would not be simple at all.  Can you elaborate?

Comment (pitrou) [inline examples/linkedlist.py]: I think you could walk the graph of types from those defined in `spec` (using the types' `__dict__` for example) and replace all deferred type instances with the actual type instances.
(ok, that might not be entirely "simple")

Comment (sklam) [inline examples/linkedlist.py]: How can we change the spec before the actual type is created?    And, what about mutual recursive types?

Comment (pitrou) [inline examples/linkedlist.py]: No, I was thinking after the type is created.

Comment (sklam) [inline examples/linkedlist.py]: I think I got it now.  Hmmm... let's see if it is all lazy enough for this

Comment (sklam) [inline examples/linkedlist.py]: Ok, I have tried implement the ``deferred(local_name)`` approach and I think it is more complicated then the current approach.  You can find my failed attempt at https://github.com/sklam/numba/compare/experiment/jitclasses...experiment/jitclasses_alt . I can get one of the example working but not the tests.  Resolving the local name later is harder than I thought.  For some reason, frame.f_locals may not have what I need.  A major source of complexity is at "resolving" the Type while they are expected to be immutable.

Comment (pitrou) [inline examples/linkedlist.py]: I see. Sorry for the suggestion then, and thanks for trying :-)

**Comment by sklam** [comment]
Oh, I just noticed that jitclasses do not like fieldnames that starts with "_" because of ``cgutils._StructProxy``

File: numba/jitclass/base.py [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+
+        self.typer.insert_function(ConstructorTemplate(self.typer))
+        self.implement_frontend(instance_type)
+        self.implement_backend(instance_type)
+
+    def implement_frontend(self, instance_type):
+        self.implement_attribute_typing(instance_type)
+
+    def implement_attribute_typing(self, instance_type):
+        class ClassAttribute(templates.AttributeTemplate):
```
Comment (pitrou) [inline numba/jitclass/base.py]: I still wonder why you can't define this at the top-level (with `key = types.ClassInstanceTyoe`).

File: numba/jitclass/boxing.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+    addr_meminfo = c.pyapi.from_native_value(types.uintp,
+                                             c.builder.ptrtoint(meminfo,
+                                                                lluintp))
+    addr_dataptr = c.pyapi.from_native_value(types.uintp,
+                                             c.builder.ptrtoint(dataptr,
+                                                                lluintp))
+
+    # XXX: relies on runtime address
+    int_addr_typ = c.context.get_constant(types.uintp, id(typ))
```
Comment (pitrou) [inline numba/jitclass/boxing.py]: The `_numba_type_` could be set in the class dict (instead of the instance's).

File: numba/jitclass/boxing.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+
+    box_subclassed = _specialize_box(typ)
+    # Note: the ``box_subclassed`` is kept alive by the cache
+    int_addr_boxcls = c.context.get_constant(types.uintp, id(box_subclassed))
```
Comment (pitrou) [inline numba/jitclass/boxing.py]: ... but of course, there's still a raw pointer constant here. I can't think of any easy solution, apart from merging Box into the MRO of the class returned by `@jitclass`.

File: numba/jitclass/decorators.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
+from __future__ import absolute_import, print_function
+from numba import types
+from .base import register_class_type, ClassBuilder
+from .immutable import ImmutableClassBuilder
+
+
+def jitclass(spec, immutable=False):
```
Comment (pitrou) [inline numba/jitclass/decorators.py]: By the way, we would probably like to pass other parameters here (at least `cache=True`, I think)...

File: numba/types.py [hunk 17]
```diff
@@ -1400,6 +1400,105 @@ class Slice3Type(Type):
     pass
 
 
+class ClassInstanceType(Type):
```
Comment (pitrou) [inline numba/types.py]: Can you add docstrings over all the new types?

File: numba/types.py [hunk 18]
```diff
@@ -1400,6 +1400,105 @@ class Slice3Type(Type):
...
+class ImmutableClassRefType(Type):
+    def __init__(self, instance_type):
+        self.instance_type = instance_type
+        super(ImmutableClassRefType, self).__init__(
+            "*{0}".format(instance_type.name))
+
+
+# Utils
+
+def is_int_tuple(x):
```
Comment (pitrou) [inline numba/types.py]: This must be a leftover from old code, it's not used anywhere AFAICT.

File: numba/runtime/nrt.c [hunk 19]
```diff
@@ -233,6 +247,18 @@ NRT_MemInfo *NRT_MemInfo_alloc_safe(size_t size) {
     return mi;
 }
 
+NRT_MemInfo* NRT_MemInfo_alloc_dtor_safe(size_t size, NRT_dtor_function dtor) {
```
Comment (pitrou) [inline numba/runtime/nrt.c]: This function can probably be unified with `NRT_MemInfo_alloc_safe`, no? It seems the `(void *) size` above should not be useful anymore...

File: numba/jitclass/decorators.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+    Args
+    ----
+    spec:
+        Specifies the types of each field on this class.
+        Must be a dictionary or a sequence.
+        With a dictionary, use collections.OrderedDict for stable ordering.
+        With a sequence, it must contain 2-tuples of (fieldname, fieldtype).
+
+    immutable:
```
Comment (pitrou) [inline numba/jitclass/decorators.py]: Since immutable isn't implemented yet, you should be sure to raise an error if set (or simply remove it from the signature).

File: numba/jitclass/base.py [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+
+
+class ClassBuilder(object):
+    """
+    A jitclass builder for mutable jitclasses.
+    """
+    instance_type_class = types.ClassInstanceType
+    per_backend_registry = {}
+
+    def __init__(self, class_type, methods, typer, backend):
```
Comment (pitrou) [inline numba/jitclass/base.py]: Just a nit, but I would rather keep calling "backend" "targetctx" or something like that. Otherwise it gets confusing.

**Comment by pitrou** [comment]
This PR lacks some docs. Perhaps you could add a new page in the user manual?

File: numba/jitclass/decorators.py (hunk 16 above)
Comment (sklam) [inline numba/jitclass/decorators.py]: Yes, we can add them later.  I don't think I should add them now, since nothing else is supported.

File: numba/jitclass/decorators.py (hunk 20 above)
Comment (sklam) [inline numba/jitclass/decorators.py]: I will disable it along with the tests at the very end

File: numba/types.py (hunk 17 above)
Comment (sklam) [inline numba/types.py]: Added in https://github.com/sklam/numba/commit/90f6a87c41fd22cc2c99b20cf7fa82ad6e55ce8b

File: docs/source/user/jitclass.rst [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+the jitclass instance is handed to the interpreter.  It is during attribute
+access to the field values that they are boxed.
+
+
+Limitations
+===========
+
+* A jitclass class object is treated as a function (the constructor) inside
+  a numba compiled function.
+* ``instance()`` only works in the interpreter.
```
Comment (pitrou) [inline docs/source/user/jitclass.rst]: isinstance()?

File: docs/source/user/jitclass.rst [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+* Manipulating jitclass instances in the interpreter is not optimized, yet.
+* Support for jitclasses are available on CPU only. (Note: support on GPU
+  devices is work-in-progress.)
```
Comment (pitrou) [inline docs/source/user/jitclass.rst]: I don't think the documentation needs to mention work-in-progress features.

File: numba/jitclass/base.py [hunk 24]
```diff
@@ -0,0 +1,415 @@
...
+
+    def implement_method(self, registry, instance_type, attr):
+        # TODO: we should be able to refactor this so that the registration
+        #       can be at the top-level instead of at per instance type.
+        @registry.register
+        @imputils.implement((instance_type, attr), types.VarArg(types.Any))
+        def imp(context, builder, sig, args):
+            method = instance_type.jitmethods[attr]
+            method.compile(sig)
+            cres = method._compileinfos[sig.args]
```
Comment (pitrou) [inline numba/jitclass/base.py]: Rather than accessing `_compileinfos`, you could stick to non-internal APIs, e.g.:
```
call = context.get_function(types.Dispatcher(method), sig)
out = call(builder, args)
```

File: numba/jitclass/base.py [hunk 25]
```diff
@@ -0,0 +1,415 @@
...
+                                                  kind='data')
+        data = data_struct(context, builder, ref=data_pointer)
+        return imputils.impl_ret_borrowed(context, builder,
+                                          typ.struct[attr],
+                                          getattr(data, attr))
+    elif attr in typ.jitprops:
+        getter = typ.jitprops[attr]['get']
+        sig = tuple([typ])
+        getter.compile(sig)
+        cres = getter._compileinfos[sig]
```
Comment (pitrou) [inline numba/jitclass/base.py]: Same here.

File: numba/jitclass/base.py (hunk 24 above)
Comment (sklam) [inline numba/jitclass/base.py]: I didn't know we could do that.  I am too used to the old way.

File: docs/source/user/jitclass.rst (hunk 23 above)
Comment (seibert) [inline docs/source/user/jitclass.rst]: I think we could phrase this as "(Note: Support for GPU devices is planned for a future release.)"  It is good to answer the obvious question in the docs if we can.  :)

**Comment by seibert** [comment]
OK, thanks everyone for the review.   Additional features or changes should go into a new PR.
